### PR TITLE
[FUN-1332] Allowlist optimisation

### PR DESCRIPTION
**Pull request by agparadiso**
### Description
this pr introduces an optimization for allowlist by updating it in batches giving priority to latest allowed addresses and adding an extra step on the batching process in which we check for the length of the allowlist and if it changed (new addresses were added) we fetch these before continuing the loop

#### What changed

We are adding an extra step that will fetch the latest addresses added IF the length of the allowlist was increased.

We were already fetching the allowlist from the contract in batches, but we were updating the node representation only when the full list was fetched. This represented a delay for users when a new address was added, which would grow proportional to the increase of the allowlist length.

In order to mitigate this, we are updating the nodes representation with every batch we fetch. 

Because the nodes representation uses a `atomic.Pointer[map[common.Address]struct{}]` we only have a `Store` method to replace completely the current list. 

So we what we are doing is:
- get the current allowlist node's state by using the `a.allowlist.Load()` 
- fetch the allowlist in batches from the contract (starting by latest addresses added)
- append the batch to a copy of the node's current state and replace the list
- clean the blocked list at the db layer

by doing this the user should be able to see the new added addresses faster than waiting until the full list is fetched.

miro diagram [here](https://miro.com/app/board/uXjVKV2M-XI=/?share_link_id=202355376485)

#### How is this tested
This is tested with white box unit tests that mock the onchain part and allows to validate the correct fetching order

### CODE REVIEW

**Review by KuphJr — COMMENTED** [review]
The general approach seems valid to me. Justs add some tests and I think we can give this another review.

File: core/services/gateway/handlers/functions/allowlist/allowlist.go [hunk 1]
```diff
@@ -254,50 +244,109 @@ func (a *onchainAllowlist) updateFromContractV1(ctx context.Context, blockNum *b
...
 	throttleTicker := time.NewTicker(time.Duration(a.config.FetchingDelayInRangeSec) * time.Second)
-	for idxStart := uint64(0); idxStart < count; idxStart += uint64(a.config.OnchainAllowlistBatchSize) {
+
+	for i := int64(currentAllowedSenderCount); i > 0; i -= int64(a.config.OnchainAllowlistBatchSize) {
 		<-throttleTicker.C
 
-		idxEnd := idxStart + uint64(a.config.OnchainAllowlistBatchSize)
-		if idxEnd >= count {
-			idxEnd = count - 1
+		idxStart := uint64(i) - uint64(a.config.OnchainAllowlistBatchSize)
```
Comment (bolekk) [inline core/services/gateway/handlers/functions/allowlist/allowlist.go]: What if count is less than batch size?

File: core/services/gateway/handlers/functions/allowlist/allowlist.go [hunk 2]
```diff
@@ -254,50 +244,109 @@ func (a *onchainAllowlist) updateFromContractV1(ctx context.Context, blockNum *b
...
+		idxEnd := uint64(i) - 1
+		if idxEnd >= currentAllowedSenderCount {
```
Comment (bolekk) [inline core/services/gateway/handlers/functions/allowlist/allowlist.go]: Is this possible?

File: core/services/gateway/handlers/functions/allowlist/allowlist.go [hunk 3]
```diff
@@ -254,50 +244,109 @@ func (a *onchainAllowlist) updateFromContractV1(ctx context.Context, blockNum *b
...
 			Context:     ctx,
-		}, idxStart, idxEnd)
+		})
 		if err != nil {
-			return nil, errors.Wrap(err, "error calling GetAllowedSendersInRange")
+			return errors.Wrap(err, "unexpected error while fetching the updated functions_allow_list.GetAllowedSendersCount")
+		}
+
+		if updatedAllowedSenderCount > currentAllowedSenderCount {
+			lastBatchIdxStart := updatedAllowedSenderCount - (updatedAllowedSenderCount - currentAllowedSenderCount)
```
Comment (bolekk) [inline core/services/gateway/handlers/functions/allowlist/allowlist.go]: this is equal to currentAllowedSenderCount

File: core/services/gateway/handlers/functions/allowlist/allowlist_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			BlockConfirmations:        1,
+			UpdateFrequencySec:        2,
+			UpdateTimeoutSec:          1,
+			StoredAllowlistBatchSize:  2,
+			OnchainAllowlistBatchSize: 10,
+			FetchingDelayInRangeSec:   1,
+		}
+
+		// allowlistSize defines how big the mocked allowlist will be
+		allowlistSize := 50
```
Comment (bolekk) [inline core/services/gateway/handlers/functions/allowlist/allowlist_internal_test.go]: Can you add a test case where this is smaller than batch size please?

File: core/services/gateway/handlers/functions/allowlist/allowlist.go (hunk 1 above)
Comment (agparadiso) [reply core/services/gateway/handlers/functions/allowlist/allowlist.go]: good catch, I was expecting uint to handle this if the subtraction was less than 0. TIL 🙇🏼

File: core/services/gateway/handlers/functions/allowlist/allowlist.go (hunk 2 above)
Comment (agparadiso) [reply core/services/gateway/handlers/functions/allowlist/allowlist.go]: no, its a leftover, thanks!

File: core/services/gateway/handlers/functions/allowlist/allowlist.go (hunk 3 above)
Comment (agparadiso) [reply core/services/gateway/handlers/functions/allowlist/allowlist.go]: true! I placed it as I thought it, but indeed:
`b - (b - a) = a`

**Review by bolekk — APPROVED** [review]

**Review by KuphJr — APPROVED** [review]

**Pull request merged**